File: BancoSangre/Controllers/PreguntasController.cs
using System;
using System.Data.Entity.Infrastructure;
using System.Linq;
using System.Web.Mvc;
using BancoSangre.Models;

namespace BancoSangre.Controllers
{
	public class PreguntasController : Controller
	{
		private readonly bancosangreEntities _db;

		public PreguntasController()
		{
			_db = new bancosangreEntities();
			_db.Configuration.ProxyCreationEnabled = false;
		}

		#region Api

		[HttpGet]
		[Authorize]
		public ActionResult ObtenerPreguntas()
		{
			var preguntas = _db.Pregunta.OrderByDescending(x => x.Mostrar).ThenBy(x => x.Orden)
				.Select(x => new
				{
					x.IdPregunta,
					x.TextoPregunta,
					x.EsTitulo,
					x.NuevaLinea,
					x.LineaCompleta,
					x.EsCerrada,
					x.CausalRechazo,
					x.Mostrar,
					x.Orden
				})
				.ToList();

			if (!preguntas.Any())
				return Json(null, JsonRequestBehavior.AllowGet);

			var result = preguntas;
			var json = new
			{
				cantidad = preguntas.Count,
				data = result
			};

			return Json(json, JsonRequestBehavior.AllowGet);
		}

		[HttpPost]
		[Authorize]
		public ActionResult RemoverPregunta(int id)
		{
			try
			{
				var actual = _db.Pregunta
					.FirstOrDefault(x => x.IdPregunta == id);
				if (actual == null)
					return Json("La pregunta que intenta remover no existe en nuestra Base de Datos.", JsonRequestBehavior.AllowGet);

				_db.Pregunta.Remove(actual);
				_db.SaveChanges();
				return Json(true, JsonRequestBehavior.AllowGet);
			}
			catch (DbUpdateException)
			{
				return Json("La pregunta no puede ser removida porque ya ha sido incluída en uno o más cuestionario/s de paciente/s.", JsonRequestBehavior.AllowGet);
			}
			catch (Exception)
			{
				return Json("Se produjo un error al intentar remover la pregunta.", JsonRequestBehavior.AllowGet);
			}
		}

		[HttpPost]
		[Authorize]
		public ActionResult GuardarPregunta(Pregunta pregunta)
		{
			try
			{
				if (pregunta.IdPregunta == 0)
				{
					_db.Pregunta.Add(pregunta);
				}
				else
				{
					var preguntaActual = _db.Pregunta.FirstOrDefault(x => x.IdPregunta == pregunta.IdPregunta);
					if (preguntaActual == null)
						throw new Exception();

					preguntaActual.TextoPregunta = pregunta.TextoPregunta;
					preguntaActual.EsTitulo = pregunta.EsTitulo;
					preguntaActual.NuevaLinea = pregunta.NuevaLinea;
					preguntaActual.LineaCompleta = pregunta.LineaCompleta;
					preguntaActual.EsCerrada = pregunta.EsCerrada;
					preguntaActual.CausalRechazo = pregunta.CausalRechazo;
					preguntaActual.Mostrar = pregunta.Mostrar;
					preguntaActual.Orden = pregunta.Orden;
				}

				_db.SaveChanges();
				return Json(true, JsonRequestBehavior.AllowGet);
			}
			catch (Exception ex)
			{
				return Json(ex.Message, JsonRequestBehavior.AllowGet);
			}
		}

		[HttpPost]
		[Authorize]
		public ActionResult ActualizarOrdenPreguntas(Pregunta preguntaMovida, Pregunta preguntaSolapa)
		{
			try
			{
				var preguntaActual = _db.Pregunta.FirstOrDefault(x => x.IdPregunta == preguntaMovida.IdPregunta);
				if (preguntaActual == null)
					throw new Exception();

				preguntaActual.Orden = preguntaMovida.Orden;

				if (preguntaSolapa != null)
				{
					var preguntaSolapada = _db.Pregunta.FirstOrDefault(x => x.IdPregunta == preguntaSolapa.IdPregunta);
					if (preguntaSolapada == null)
						throw new Exception();

					preguntaSolapada.Orden = preguntaSolapa.Orden;
				}

				_db.SaveChanges();
				return Json(true, JsonRequestBehavior.AllowGet);
			}
			catch (Exception ex)
			{
				return Json(ex.Message, JsonRequestBehavior.AllowGet);
			}
		}
		
		#endregion

		// GET: Preguntas
		[Authorize]
		public ActionResult Index()
		{
			return View();
		}
		
		protected override void Dispose(bool disposing)
		{
			if (disposing)
			{
				_db.Dispose();
			}
			base.Dispose(disposing);
		}
	}
}

File: BancoSangre/app/controllers/preguntasController.js
"use strict";

app.controller("preguntasController", function ($scope, preguntasRepositorio, modalServicio) {
	init();

	function init() {
		configCheckMostrar();
		configPaginacion();
		obtenerPreguntas();
	}

	function configPaginacion() {
		$scope.infoPagina = {
			pagina: 1,
			itemsPorPagina: 9,
			reversa: false,
			totalItems: 0
		};
	}

	function obtenerPreguntas() {
		preguntasRepositorio.obtenerPreguntas()
			.then(function (result) {
				$scope.preguntas = result.data !== "" ? result.data.data : [];
				$scope.infoPagina.totalItems = result.data.cantidad;
			});
	}

	function configCheckMostrar() {
		$("#checkMostrar").prop("indeterminate", true).data("checked", 0).on("click",
			function () {
				var el = $(this);

				switch (el.data("checked")) {
					case 0:
						el.data("checked", 1);
						el.prop("indeterminate", false);
						el.prop("checked", true);
						break;
					case 1:
						el.data("checked", 2);
						el.prop("indeterminate", false);
						el.prop("checked", false);
						break;
					default:
						el.data("checked", 0);
						el.prop("indeterminate", true);
						el.prop("checked", false);
				}
			});
	}

	$scope.removerPregunta = function (id) {
		preguntasRepositorio.remover(id)
			.then(function (result) {
				if (result.data.toString() === "true") {
					var i;
					for (i = 0; i < $scope.preguntas.length; ++i) {
						if ($scope.preguntas[i].IdPregunta === id) {
							$scope.preguntas.splice(i, 1);
							break;
						}
					}
					modalServicio.open("success", "La pregunta ha sido removida correctamente.");
				}
				else {
					modalServicio.open("danger", result.data);
				}
			});
	};

	$scope.cancelarPregunta = function (rowform, id) {
		if (id <= 0) {
			var i;
			for (i = 0; i < $scope.preguntas.length; ++i) {
				if ($scope.preguntas[i].IdPregunta === id) {
					$scope.preguntas.splice(i, 1);
					break;
				}
			}
		}
		rowform.$cancel();
	}

	$scope.agregarPregunta = function () {
		$scope.inserted = {
			IdPregunta: 0,
			TextoPregunta: "",
			EsTitulo: false,
			NuevaLinea: false,
			LineaCompleta: false,
			EsCerrada: false,
			CausalRechazo: false,
			Mostrar: false,
			Orden: $scope.preguntas.length + 1
		};
		$scope.preguntas.push($scope.inserted);
	};

	function crearPregunta(id, textoPregunta, esTitulo, nuevaLinea, lineaCompleta, esCerrada, causalRechazo, mostrar, orden) {
		return {
			IdPregunta: id,
			TextoPregunta: textoPregunta,
			NuevaLinea: nuevaLinea,
			EsTitulo: esTitulo,
			LineaCompleta: lineaCompleta,
			EsCerrada: esCerrada,
			CausalRechazo: causalRechazo,
			Mostrar: mostrar,
			Orden: orden
		}
	}

	$scope.guardarPregunta = function (data, id) {
		var pregunta = crearPregunta(id, data.textoPregunta, data.esTitulo, data.nuevaLinea, data.lineaCompleta, data.esCerrada, data.causalRechazo, data.mostrar, data.orden);
		preguntasRepositorio.guardar(pregunta)
			.then(function (result) {
				if (result.data) {
					obtenerPreguntas();
					modalServicio.open("success", "La pregunta se ha guardado con éxito.");
				}
				else {
					modalServicio.open("danger", "Error al guardar la pregunta.");
				}
			});
	};

	function actualizarOrden(preguntaMovida, preguntaSolapa) {
		if (preguntaMovida != null) {
			preguntasRepositorio.actualizarOrden(preguntaMovida, preguntaSolapa)
				.then(function (result) {
					if (result) {
						obtenerPreguntas();
					} else {
						modalServicio.open("danger", "Error al actualizar el orden.");
					}
				});
		}
	}

    $scope.cambiarOrden = function (id, ordenActual) {
        if (ordenActual <= 0) return;

		var i;
		var preguntaMovida = null;
		var preguntaSolapa = null;
		var ordenParaSolapada = 0;
		for (i = 0; i < $scope.preguntas.length; ++i) {
			if ($scope.preguntas[i].IdPregunta === id) {
				preguntaMovida = $scope.preguntas[i];
				ordenParaSolapada = preguntaMovida.Orden;
				preguntaMovida.Orden = ordenActual;
				break;
			}
		}
		for (i = 0; i < $scope.preguntas.length; ++i) {
			if ($scope.preguntas[i].Orden === ordenActual && $scope.preguntas[i].IdPregunta !== id) {
				preguntaSolapa = $scope.preguntas[i];
				preguntaSolapa.Orden = ordenParaSolapada;
				break;
			}
		}

		actualizarOrden(preguntaMovida, preguntaSolapa);
	}

	$scope.filtrar = function () {
		// Siempre filtra por el texto, pero además, si el check de mostrar no está en estado indeterminado también filtramos por ese check.
		var checkMostrarEnIndeterminado = $("#checkMostrar").prop("indeterminate");
		if (!checkMostrarEnIndeterminado) {
			$scope.filtro = { "TextoPregunta": $("#busqTextoPregunta").val(), "Mostrar": $("#checkMostrar").prop("checked") };
		} else {
			$scope.filtro = { "TextoPregunta": $("#busqTextoPregunta").val() };
		}
	};

});